Modules/_ctypes/callproc.c

```diff
@@ -1287,11 +1288,26 @@ static PyObject *load_library(PyObject *self, PyObject *args)
         return NULL;
 
     Py_BEGIN_ALLOW_THREADS
-    hMod = LoadLibraryW(name);
+    /* bpo-36085: Limit DLL search directories to avoid pre-loading
+     * attacks and enable use of the AddDllDirectory function.
+     */
+    hMod = LoadLibraryExW(name, NULL,
+                          LOAD_LIBRARY_SEARCH_DEFAULT_DIRS |
+                          LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR);
```

Comment:
We cannot use `LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR` as a default in ctypes. It will fail the call as an invalid parameter if  `name` isn't fully qualified. This flag doesn't even allow paths that are explicitly relative to the working directory due to the way path searching works in Windows. For example, ".\spam\eggs.dll" is joined with every directory in the computed DLL search path (e.g. "C:\Windows\System32\.\spam\eggs.dll"). The secure search path never includes the current working directoryand, so the loader can't assume we want the given path to be resolved relative to the working directory in order to compute the DLL directory. It has to fail the call. (In contrast, with POSIX `dlopen` "./spam/eggs.so" is only resolved relative to the working directory.)

Reply:
Yes, you're right., I'll remove it again.

Making the second argument consistent with dlopen is a much bigger task that I can't afford to take on right now, so I'm going to leave that open. We can discuss on the bug whether that should block the entire change or not.

Reply:
Do you think something should be done about the relative-path issue? The behavior is documented:

> If lpFileName specifies a relative path, the entire relative path is appended to every token in the DLL search path. To load a module from a relative path without searching any other path, use GetFullPathName to get a nonrelative path and call LoadLibraryEx with the nonrelative path. 

IMO, this behavior is counter-intuitive and disagrees with POSIX. I'd expect `CDLL("spam/eggs.dll")` to resolve the path relative to the working directory. I don't think most people would intuit that the system actually searches relative to each directory in the computed DLL search path. This seems to work with the expected behavior when the current directory is in the DLL search path. Albeit, with the default safe order, the working directory is checked after the application, System32, and Windows directories, but a collision with a system DLL is unlikely. On the other hand, with the secure search path, which doesn't include the working directory, scripts will almost definitely fail if they depend on an explicit reference to the working directory (e.g. "./eggs.dll" or "spam/eggs.dll"). 

If we were Windows only, I'd point people to the documentation and tell them to adjust their expectations appropriately. But we're cross-platform, and the Windows behavior is the odd duck here. Maybe we should call `GetFullPathNameW` on any path that includes slash or backslash in order to match expected behavior.

Reply:
Alternatively, ctypes could fail the call as a `ValueError` if the name contains slash or backslash and isn't fully qualified. This avoids the unexpected search and enforces a policy that ctypes in Windows will no longer search the working directory for unqualified names nor load DLLs relative to the working directory. The downside with this approach is that it's inconsistent with POSIX.

Note that `os.add_dll_directory` won't allow adding ".", so there's no supported way for users to sneak the old behavior back in anyway. The working directory can be restored to the search path by using ctypes to call `SetDllDirectoryW(".")`, but that's beyond our control.